tests.py

```diff
@@ -1,129 +1,523 @@
+    But a TextIOWrapper around self.rfile would buffer too many bytes
+    from the stream, bytes which we later need to read as bytes.
+    So we read the correct bytes here, as bytes, for email Parser
+    to parse.
+    """
+    headers = []
+
+    while True:
+        line = fp.readline(_MAXLINE + 1)
+
+        if len(line) > _MAXLINE:
+            raise Exception("LineTooLong: header line")
+        headers.append(line)
+        if len(headers) > _MAXHEADERS:
+            raise Exception("HTTPException: got more than %d headers" % _MAXHEADERS)
+        if line in (b"\r\n", b"\n", b""):
+            break
+    header_bytes = b"".join(headers)
+    hstring = header_bytes.decode("iso-8859-1")
+    return email.parser.Parser(_class=_class).parsestr(hstring), header_bytes
+
+
+def get_meth_text_pos(text, method):
+    """ searches for line: `<code> #<method>-<shift>`
+          - example: `sys.getdefaultencoding() #{METHOD_HOVER}-5`
+            position returned will be 5 chars before `#...`: `sys.getdefaultencodi | ng() `
+        returns (x,y)
+    """
+    meth_mark = "#" + method
+    lines = text.splitlines()
+
+    # line index
+    target_line_ind = next(i for i, line in enumerate(lines) if meth_mark in line)
+    # char index
+    mark_character_ind = lines[target_line_ind].index(meth_mark)
+    m = re.search(f"\\#{method}-(\\d+)", text)
+    target_character_ind = mark_character_ind - int(m.group(1))
+
+    return (target_character_ind, target_line_ind)
+
+
+class ThreadedServer:
+    """ Gathers all messages received from server - to handle random-order-messages \
+            that are not a response to a request.
+
+        * get_msg_by_type() - get server message by type. waits for the message
+        * stop()            - stop server via LSP
+    """
+
+    def __init__(self, process, root_uri):
+        self.process = process
+        self.root_uri = root_uri
+        self.lsp_client = lsp.Client(
+            root_uri=root_uri,
+            workspace_folders=[lsp.WorkspaceFolder(uri=self.root_uri, name="Root")],
+            trace="verbose",
+        )
+        self.msgs = []
+
+        self._pout = process.stdout
+        self._pin = process.stdin
+
+        self._read_q = queue.Queue()
+        self._send_q = queue.Queue()
+
+        self.reader_thread = threading.Thread(
+            target=self._read_loop, name="lsp-reader", daemon=True
+        )
+        self.writer_thread = threading.Thread(
+            target=self._send_loop, name="lsp-writer", daemon=True
+        )
+
+        self.reader_thread.start()
+        self.writer_thread.start()
+
+        self.exception = None
+
+    @property
+    def all_msgs(self):
+        return self.msgs[:]
+
+    # threaded
+    def _read_loop(self):
+        try:
+            while self._pout:
+                headers, header_bytes = parse_headers(self._pout)  # type: ignore
+
+                LOG_IN and print(f"\n <<< received: {headers}\n")
```

Comment:
Nit: while this works, an explicit `if` statement would be preferred.

Reply:
Ok, will change